Documentation/guides/OneDotNetEmbeddedResources.md

```diff
@@ -0,0 +1,288 @@
+[assembly: Java.Interop.Dependency ("bar.aar")]
+```
+
+We can emit a build error if any files are missing.
+
+We also need to emit an attribute to mark the assembly as using the
+new system:
+
+```csharp
+[assembly: Java.Interop.Features (Java.Interop.Feature.NativeDependencies)]
+```
+
+`Feature.NativeDependencies` would allow us to skip the assembly
+during the `<ResolveLibraryProjectImports/>` MSBuild task. We would
+need to emit this for every Xamarin.Android class library, as new
+libraries might not contain any `[assembly: Dependency]` attributes.
+
+Usage of an enum would allow us to add new things later:
+
+```csharp
+public enum Feature
+{
+    NativeDependencies   = 0x1,
+    // Possible new features
+    // AndroidManifest   = 0x2
+    // JavaStubs         = 0x4
+    // MavenDependencies = 0x8,
+}
+```
+
+The above attributes will be auto-generated at build time in
+`$(IntermediateOutputPath)__Features__.cs`, developers will not
+intentionally add these attributes to their projects. We should
+potentially add `[EditorBrowsable (EditorBrowsableState.Never)]`, so
+they are not visible to Intellisense.
+
+### MSBuild Item Groups
+
+Right now we have a confusing collection of MSBuild item groups:
+
+* `@(InputJar)` - `.jar` file to create a Java binding for, but not
+  embed.
+* `@(EmbeddedJar)` - `.jar` file to create a Java binding and embed.
+* `@(ReferenceJar)` - `.jar` file to reference for the binding
+  process.
+* `@(EmbeddedReferenceJar)` - `.jar` file to reference and embed, but
+  not bind.
+* `@(LibraryProjectZip)` - `.aar` or `.zip` file to bind and embed.
+  Only one can be used in a project!
+* `@(AndroidAarLibrary)` - `.aar` file to include in a application
+  project.
+* `@(AndroidJavaLibrary)` - `.jar` file to include in a application
+  project.
+* `@(EmbeddedNativeLibrary)` - `.so` file to embed in a class library
+  project or include in an application project.
+* `@(AndroidNativeLibrary)` - `.so` file to include in an application
+  project.
+
+Let's simplify this, we could support all of the above with a new
+`@(AndroidLibrary)` item group:
+
+```xml
+<!-- Include and bind -->
+<AndroidLibrary Include="foo.aar" Bind="true" />
+<!-- Just include, do not bind -->
+<AndroidLibrary Include="bar.aar" />
+<AndroidLibrary Include="baz.jar" />
+<!--
+  Bind but do not include in NuGet package.
+  %(Pack) is built into NuGet MSBuild targets.
+-->
+<AndroidLibrary Include="bar.aar" Bind="true" Pack="false" />
+<!-- Native libraries need ABI directory -->
+<AndroidLibrary Include="armeabi-v7a\libfoo.so" />
+<AndroidLibrary Include="x86\libfoo.so" />
+```
+
+The new `@(AndroidLibrary)` item group will simply translate to the
+old ones for backwards compatibility. The extension of the file can be
+used to determine what kind of library each item is. `%(Bind)` will be
+`false` by default, and `%(Pack)` will be `true` by default. `%(Pack)`
+will not do anything in application projects.
```

Comment:
@Redth mentioned we probably need a way to exclude a `.jar` or `.aar` file in an application project.

There are cases, with like Amazon Kindle Fire SDKs or Symbol/Zebra EMDK, you might want C# bindings for the library, but the actual device has the library installed system wide. In this case, you should not include the code in the `classes.dex` in your app.

In that case would `Pack="false"` make sense, if you for some reason want to exclude a `.jar`, `.aar`, `.so` from your `.apk` file?

Reply:
How are those files *currently* excluded?

We're going to need to have a migration path from current projects to .NET6+ projects, and the easier this migration can be done, the better.  Presumably if we can reuse the current exclusion mechanism, it'll simplify migration.

Reply:
Right now if you wanted to distribute a NuGet package of a binding like this, you would have to:
1. Use `@(InputJar)` for the binding.
2. Add a `.targets` file the NuGet package that includes the `.jar` file as `@(AndroidExternalJavaLibrary)` for application projects.

Maybe we just leave it as-is, I don't see a good way to do better at the moment. Checking for `%(Pack)`= `false` would only help when adding `.aar`/`.jar` files directly to an application project. It doesn't help NuGet packages, as you still need an MSBuild `.targets` file.

Reply:
I don't think we can leave it "as-is" -- I think we want to *drop* the `@(InputJar)` name? -- but I think it would be useful to be *similar* to the current state of affairs, and "repurpose" `@(AndroidExternalJavaLibrary)` as a "synonym" for `@(Inputjar)` within projects.

Or perhaps go with `@(AndroidExternalLibrary)` to "mirror" `@(AndroidLibrary)`?.

Otherwise we have to go with `%(AndroidLibrary.Pack)`=False, which is also entirely "sensible" and "workable", but I'm less certain about *tooling support*.  We might need @joj to chime in on what would be easiest for the IDE to support.